File: results/0.75/DNS/dns-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <hammer/glue.h>
#include <hammer/parsers/parser_internal.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct {
    uint16_t id;
    uint8_t qr;
    uint8_t opcode;
    uint8_t aa;
    uint8_t tc;
    uint8_t rd;
    uint8_t ra;
    uint8_t z;
    uint8_t rcode;
    uint16_t qdcount;
    uint16_t ancount;
    uint16_t nscount;
    uint16_t arcount;
} DNSHeader;

typedef struct {
    char* qname;
    uint16_t qtype;
    uint16_t qclass;
} DNSQuestion;

typedef struct {
    char* name;
    uint16_t type;
    uint16_t class;
    uint32_t ttl;
    uint16_t rdlength;
    uint8_t* rdata;
} DNSResourceRecord;

static HParser* dns_header_parser;
static HParser* dns_question_parser;
static HParser* dns_rr_parser;
static HParser* dns_message_parser;

static HParsedToken* act_make_header(const HParseResult* p, void* user_data) {
    DNSHeader* header = (DNSHeader*)malloc(sizeof(DNSHeader));
    const HCountedArray* seq = H_CAST_SEQ(p->ast);
    uint16_t flags = H_CAST_UINT(seq->elements[1]);
    
    header->id = H_CAST_UINT(seq->elements[0]);
    header->qr = (flags >> 15) & 0x1;
    header->opcode = (flags >> 11) & 0xF;
    header->aa = (flags >> 10) & 0x1;
    header->tc = (flags >> 9) & 0x1;
    header->rd = (flags >> 8) & 0x1;
    header->ra = (flags >> 7) & 0x1;
    header->z = (flags >> 4) & 0x7;
    header->rcode = flags & 0xF;
    header->qdcount = H_CAST_UINT(seq->elements[2]);
    header->ancount = H_CAST_UINT(seq->elements[3]);
    header->nscount = H_CAST_UINT(seq->elements[4]);
    header->arcount = H_CAST_UINT(seq->elements[5]);
    
    HArena* arena = h_new_arena();
    HParsedToken* ret = h_make_seqn(arena, 1);
    ret->seq->elements[0] = (HParsedToken*)header;
    return ret;
}

static HParser* init_dns_header_parser(void) {
    return h_action(h_sequence(h_uint16(), h_uint16(),
                              h_uint16(), h_uint16(),
                              h_uint16(), h_uint16(),
                              NULL),
                    act_make_header, NULL);
}

static HParser* init_dns_label_parser(void) {
    return h_sequence(h_uint8(),
                     h_repeat_n(h_uint8(), h_uint8()),
                     NULL);
}

static HParser* init_dns_name_parser(void) {
    return h_many1(init_dns_label_parser());
}

static HParser* init_dns_question_parser(void) {
    return h_sequence(init_dns_name_parser(),
                     h_uint16(),
                     h_uint16(),
                     NULL);
}

static HParser* init_dns_rdata_parser(void) {
    return h_repeat_n(h_uint8(), h_uint16());
}

static HParser* init_dns_rr_parser(void) {
    return h_sequence(init_dns_name_parser(),
                     h_uint16(),
                     h_uint16(),
                     h_uint32(),
                     init_dns_rdata_parser(),
                     NULL);
}

static HParser* init_dns_message_parser(void) {
    return h_sequence(init_dns_header_parser(),
                     h_many(init_dns_question_parser()),
                     h_many(init_dns_rr_parser()),
                     h_many(init_dns_rr_parser()),
                     h_many(init_dns_rr_parser()),
                     NULL);
}

void init_parsers(void) {
    dns_header_parser = init_dns_header_parser();
    dns_question_parser = init_dns_question_parser();
    dns_rr_parser = init_dns_rr_parser();
    dns_message_parser = init_dns_message_parser();
}

int main(int argc, char** argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <dns_binary_file>\n", argv[0]);
        return 1;
    }

    FILE* f = fopen(argv[1], "rb");
    if (!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t* buffer = malloc(size);
    if (fread(buffer, 1, size, f) != size) {
        perror("Failed to read file");
        fclose(f);
        free(buffer);
        return 1;
    }
    fclose(f);

    init_parsers();

    HParseResult* result = h_parse(dns_message_parser, buffer, size);
    if (!result) {
        fprintf(stderr, "Failed to parse DNS message\n");
        free(buffer);
        return 1;
    }

    printf("Successfully parsed DNS message\n");

    h_parse_result_free(result);
    free(buffer);
    return 0;
}

File: results/0.75/DNS/dns-claude-3-5-sonnet-20241022.rs
use nom::{
    bytes::complete::{take, take_while},
    combinator::{map, verify},
    multi::{count, length_count},
    number::complete::{be_u16, be_u32, be_u8},
    sequence::tuple,
    IResult,
};
use std::{env, fs::File, io::Read};

#[derive(Debug)]
struct DnsHeader {
    id: u16,
    qr: bool,
    opcode: u8,
    aa: bool,
    tc: bool,
    rd: bool,
    ra: bool,
    z: u8,
    rcode: u8,
    qdcount: u16,
    ancount: u16,
    nscount: u16,
    arcount: u16,
}

#[derive(Debug)]
struct DnsQuestion {
    qname: Vec<String>,
    qtype: u16,
    qclass: u16,
}

#[derive(Debug)]
struct DnsResourceRecord {
    name: Vec<String>,
    type_: u16,
    class: u16,
    ttl: u32,
    rdlength: u16,
    rdata: Vec<u8>,
}

#[derive(Debug)]
struct DnsMessage {
    header: DnsHeader,
    questions: Vec<DnsQuestion>,
    answers: Vec<DnsResourceRecord>,
    authorities: Vec<DnsResourceRecord>,
    additionals: Vec<DnsResourceRecord>,
}

fn parse_label(input: &[u8]) -> IResult<&[u8], String> {
    let (input, length) = be_u8(input)?;
    let (input, label) = take(length)(input)?;
    Ok((input, String::from_utf8_lossy(label).into_owned()))
}

fn parse_name(input: &[u8]) -> IResult<&[u8], Vec<String>> {
    let mut labels = Vec::new();
    let mut remaining = input;

    loop {
        let (input, length) = be_u8(remaining)?;
        if length == 0 {
            return Ok((input, labels));
        }
        let (input, label) = take(length)(input)?;
        labels.push(String::from_utf8_lossy(label).into_owned());
        remaining = input;
    }
}

fn parse_header(input: &[u8]) -> IResult<&[u8], DnsHeader> {
    let (input, (id, flags, qdcount, ancount, nscount, arcount)) =
        tuple((be_u16, be_u16, be_u16, be_u16, be_u16, be_u16))(input)?;

    Ok((
        input,
        DnsHeader {
            id,
            qr: (flags & 0x8000) != 0,
            opcode: ((flags & 0x7800) >> 11) as u8,
            aa: (flags & 0x0400) != 0,
            tc: (flags & 0x0200) != 0,
            rd: (flags & 0x0100) != 0,
            ra: (flags & 0x0080) != 0,
            z: ((flags & 0x0070) >> 4) as u8,
            rcode: (flags & 0x000F) as u8,
            qdcount,
            ancount,
            nscount,
            arcount,
        },
    ))
}

fn parse_question(input: &[u8]) -> IResult<&[u8], DnsQuestion> {
    let (input, qname) = parse_name(input)?;
    let (input, (qtype, qclass)) = tuple((be_u16, be_u16))(input)?;

    Ok((
        input,
        DnsQuestion {
            qname,
            qtype,
            qclass,
        },
    ))
}

fn parse_resource_record(input: &[u8]) -> IResult<&[u8], DnsResourceRecord> {
    let (input, name) = parse_name(input)?;
    let (input, (type_, class, ttl, rdlength)) =
        tuple((be_u16, be_u16, be_u32, be_u16))(input)?;
    let (input, rdata) = take(rdlength)(input)?;

    Ok((
        input,
        DnsResourceRecord {
            name,
            type_,
            class,
            ttl,
            rdlength,
            rdata: rdata.to_vec(),
        },
    ))
}

fn parse_dns_message(input: &[u8]) -> IResult<&[u8], DnsMessage> {
    let (input, header) = parse_header(input)?;
    let (input, questions) = count(parse_question, header.qdcount as usize)(input)?;
    let (input, answers) = count(parse_resource_record, header.ancount as usize)(input)?;
    let (input, authorities) = count(parse_resource_record, header.nscount as usize)(input)?;
    let (input, additionals) = count(parse_resource_record, header.arcount as usize)(input)?;

    Ok((
        input,
        DnsMessage {
            header,
            questions,
            answers,
            authorities,
            additionals,
        },
    ))
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <dns_binary_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_dns_message(&buffer) {
        Ok((remaining, message)) => {
            println!("Parsed DNS message: {:#?}", message);
            if !remaining.is_empty() {
                println!("Warning: {} bytes remaining", remaining.len());
            }
        }
        Err(e) => eprintln!("Failed to parse DNS message: {:?}", e),
    }

    Ok(())
}